code/WorkInProgress/nadir_antenna.dm

```diff
@@ -0,0 +1,688 @@
+	proc/check_transceive()
+		. = "ERR_NO_ARRAY"
+		if(!transception_array)
+			return
+		var/datum/powernet/powernet = src.get_direct_powernet()
+		if(!powernet)
+			return "NO_WIRE_ENDPOINT"
+		var/netnum = powernet.number
+		var/error_code = transception_array.can_transceive(netnum)
+		switch(error_code)
+			if(TRANSCEIVE_BUSY) //connection's fine it's just busy at this particular time
+				return "OK"
+			if(TRANSCEIVE_NOPOWER)
+				return "ERR_ARRAY_APC"
+			if(TRANSCEIVE_POWERWARN)
+				return "ARRAY_POWER_LOW"
+			if(TRANSCEIVE_NOWIRE)
+				return "ERR_WIRE"
+			if(TRANSCEIVE_OK)
+				return "OK"
+			else
+				return "ERR_OTHER" //what
+
+	///Attempts to perform a transception operation; receive if it was passed an index for pending inbound cargo, send otherwise
+	proc/attempt_transceive(var/cargo_index = null)
+		if(src.is_transceiving)
+			return
+		if(!transception_array)
+			return
+		var/datum/powernet/powernet = src.get_direct_powernet()
+		if(!powernet)
+			return
+		var/netnum = powernet.number
+		if(transception_array.can_transceive(netnum) != TRANSCEIVE_OK)
+			return
+		if(cargo_index != null)
+			if(shippingmarket.pending_crates[cargo_index])
+				var/obj/inbound_target = shippingmarket.pending_crates[cargo_index]
+				receive_a_thing(netnum,inbound_target)
+			else
+				return
+		else
+			send_a_thing(netnum)
+		return
+
+	proc/send_a_thing(var/netnumber)
+		src.is_transceiving = TRUE
+		playsound(src.loc, "sound/effects/ship_alert_minor.ogg", 50, 0) //outgoing cargo warning (stand clear)
+		SPAWN(2 SECONDS)
+			flick("neopad_activate",src)
+			SPAWN(0.3 SECONDS)
+				var/atom/movable/thing2send
+				var/list/oofed_nerds = list()
+				for(var/atom/movable/O as obj|mob in src.loc)
+					if(O.anchored) continue
+					if(O == src) continue
+					if(istype(O,/mob)) //no mobs
+						if(istype(O,/mob/living/carbon/human) && prob(25))
+							oofed_nerds += O
+						continue
+					if(istype(O,/obj/storage/crate) || istype(O,/obj/artifact) || istype(O,/obj/machinery/artifact))
+						thing2send = O
+						break //only one thing at a time!
+				for(var/nerd in oofed_nerds)
+					telefrag(nerd) //did I mention NO MOBS
+				if(thing2send && transception_array.transceive(netnumber))
+					thing2send.loc = src
+					SPAWN(1 SECOND)
+
+						if (istype(thing2send, /obj/storage/crate/biohazard/cdc))
+							QM_CDC.receive_pathogen_samples(thing2send)
+
+						else if(istype(thing2send,/obj/storage/crate) || istype(thing2send,/obj/storage/secure/crate))
+							var/sold_to_trader = FALSE
+							for (var/datum/trader/T in shippingmarket.active_traders)
+								if (T.crate_tag == thing2send.delivery_destination)
+									shippingmarket.sell_crate(thing2send, T.goods_buy)
+									sold_to_trader = TRUE
+									break
+							if(!sold_to_trader)
+								shippingmarket.sell_crate(thing2send)
+
+						else if(istype(thing2send,/obj/artifact))
+							shippingmarket.sell_artifact(thing2send)
+
+						else //how even
```

Comment:
I don't think this works? Not all artifacts are /obj/artifact, you'd need to check for the artifact var on /obj or whatever it was, check the sell_artifact proc maybe

Reply:
That sounds like probably a good thing to do, will check into that in a bit.

Reply:
Done in feb07ad